src/stockpile.clj

```diff
@@ -0,0 +1,364 @@
+                       (pr-str (-> (:directory q) .toAbsolutePath str))))))
+     (let [tmp-dest (create-tmp-file (.resolve (:directory q) "q"))
+           encode-meta (:encode-meta q)]
+       ;; It might be possible to optimize some cases with
+       ;; transferFrom/transferTo eventually.
+       (try
+         (Files/copy stream tmp-dest
+                     (into-array [StandardCopyOption/REPLACE_EXISTING]))
+         (fsync tmp-dest false)
+         (loop [id likely-id
+                meta-retry 0
+                meta metadata
+                enc-meta (when metadata (encode-meta metadata))]
+           (let [target (-> (:directory q)
+                            (.resolve "q")
+                            (.resolve (str id (when enc-meta
+                                                (str "-" enc-meta)))))
+                 ;; Can't recur from catch
+                 rename (try
+                          (rename-durably tmp-dest target true)
+                          :committed
+                          (catch FileAlreadyExistsException ex
+                            :collision)
+                          (catch FileSystemException ex
+                            ;; or just the more general IOException?
+                            ;; i.e. openjdk-7 throws this, but can we
+                            ;; expect that everywhere for "too long"?
+                            ;; If you change this, change instance?
+                            ;; below.
+                            ex))]
+             (case rename
+               :committed
+               (do
+                 (.getAndIncrement (:next-likely-id q))
+                 (entry id meta))
+
+               :collision
+               (recur (inc id) meta-retry meta enc-meta)
+
+               (if-not (instance? FileSystemException rename)
+                 (assert false)
+                 (if-not adjust-metadata
+                   (throw
+                    (ex-info "Final commit failed" {:kind ::commit-failed}
+                             rename))
+                   (let [retry (inc meta-retry)
+                         next-meta (adjust-metadata retry meta)]
+                     (when-not next-meta
+                       (throw
+                        (ex-info "Final commit failed" {:kind ::commit-failed}
+                                 rename)))
+                     (recur id retry next-meta (encode-meta next-meta))))))))
+         (catch Exception ex
+           (Files/deleteIfExists tmp-dest)
+           (throw ex)))))))
+
+(defn- entry-path [q entry]
+  (-> (:directory q)
+      (.resolve "q")
+      (.resolve (str (entry-id entry)
+                     (when-let [m (entry-meta entry)]
+                       (when-let [em ((:encode-meta q) m)]
+                         (str "-" em)))))))
+
+(defn stream [q entry]
+  "Returns an unbuffered stream of the entry's data."
+  (Files/newInputStream (entry-path q entry) (into-array OpenOption [])))
```

Comment:
Discussed this with Ryan, and I think we'll keep it - I'd intended to leave any buffering up to the caller.